rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp

```diff
@@ -99,6 +99,18 @@ MultiThreadedExecutor::run(size_t this_thread_number)
 
     execute_any_executable(any_exec);
 
+    if (any_exec.callback_group &&
+      any_exec.callback_group->type() == CallbackGroupType::MutuallyExclusive)
+    {
+      try {
+        interrupt_guard_condition_->trigger();
```

Comment:
@mjcarroll To answer my own question, this should be the culprit of our performance regression. This will trigger a full rebuild of the collection.

Reply:
> Yes, but in this case, my understanding is that it is the correct thing to do.

Yes and no. You don't need to recollect all entities, this is super expensive. You only need to make sure, that you don't add entities that belong to a blocked callback group to the rcl_wait_set_t. 

E.g. you should have a rclcpp::WaitSet per callback group, and be able to add them sequentially

> Previously, this was done in the Executor::execute_any_executable method unconditionally 

Yep, but previously this operation also was 3 times faster :-(

Reply:
> You only need to make sure, that you don't add entities that belong to a blocked callback group to the rcl_wait_set_t.

But you also have to remove them from the waitset.  If, for instance, imagine you had a mutually exclusive group with two threads.  If you had a timer with an extremely long task (multiples of times longer than the period), but didn't remove other masked entities from the waitset, then the waitset would continuously "wake" but without any work to do.

This is a bit of a contrived example (hopefully?) but in the current implementation, any entity that isn't eligible to execute should also not be waited on.

Reply:
> But you also have to remove them from the waitset.

Now I am confused, the rcl_wait_set_t or the rclcpp::Waitset ?

>  If, for instance, imagine you had a mutually exclusive group with two threads. If you had a timer with an extremely long task (multiples of times longer than the period), but didn't remove other masked entities from the waitset, then the waitset would continuously "wake" but without any work to do.

You seem to have misunderstood what I meant, or in my mind stuff is different structured ;-)
In my mind a rclcpp::Waitset is a collection of entities, that can be waited on. In the moment, that you actually want to perform the wait, you need to build an rcl_wait_set_t from it.

This is the point, were you can decide, which entities you actually want to add to the rcl_wait_set_t and really wait on. This would be the point were you don't add entities, that are in a blocked callback group.

However, if you go down that road, you will hit the problem, that your backmapping from rcl_wait_set_t to rclcpp::Waitset is not broken, as there is no one to one relationship between the entities in the rclcpp::Waitset and rcl_wait_set_t any more. I ran into the same problem in the CBG Executor. As a solution, I added a data ptr to the entities in rcl, and set them before the wait to a variant containing a reference to my rclcpp entitiy.

Reply:
> In my mind a rclcpp::Waitset is a collection of entities, that can be waited on. In the moment, that you actually want to perform the wait, you need to build an rcl_wait_set_t from it.

Almost.  Since building the `rcl_wait_set_t` is somewhat expensive (as we have discussed), it is only built from the `rclcpp::WaitSet` in the case that the to-be-waited-on entities change.  In the case that you had each entity in a separate callback group or all entities in `Reentrant` callback groups, this would make it such that the `rcl_wait_set_t` structure would never have to be rebuilt.

> This is the point, were you can decide, which entities you actually want to add to the rcl_wait_set_t and really wait on. This would be the point were you don't add entities, that are in a blocked callback group.

That's _sort of_ what happens, except that it's done by adding/removing entities from the rclcpp::WaitSet.  The rclcpp::WaitSet has no concept of callback groups.

> However, if you go down that road, you will hit the problem, 

Every road has it's own set of problems here :).  The bookkeeping for the rclcpp::WaitSet does currently seem to be a fairly expensive part, maybe I can work at reducing that some for better performance.

Reply:
Sorry, you are correct.  It prevents the `rcl_wait_set_t` from being resized, but it does have to be re-initialized before each wait.

Reply:
> The rclcpp::WaitSet has no concept of callback groups.

I think this might be the most important point @mjcarroll made. It's up to the executor to manipulate the wait set as needed to control what is waited on according to constraints by the callback groups.

In the past, however, I had suggested a new concept for the wait set (at the C++ level and all the way down through rcl and rmw too) which I was calling a mask. That way you can have a wait set with a few things in it and then periodically you can wait on it with a mask to avoid waiting on a few of them (temporarily). That would allow for much better optimizations at the rmw layer, where this is most expensive. Resetting the contents of the rcl wait set is not that expensive, it's just lists of pointers. The expensive part is usually at the DDS layer where there is a somewhat complex activity of "attaching" and "detaching" conditions to their wait set object.

The other expensive part (which we may be able to avoid) is (re)building the collection of rclcpp things. We may still need to build or modify an existing wait set from the collection, but maybe we could save some time there. We still have to handle the asynchronous destruction of entities though, which is something we have to check about the collection each time we use it, which is part of the expensive operations (essentially locking the weak pointers to make sure they're still valid). But that is just speculation on my part, more likely we have to keep doing what we're doing, but make it more efficient instead.

Reply:
To be clear, the mask idea is out-of-scope for this pr, but I think it strikes at the root of the issue. I still want to look at why the multi-threaded is so much slower, but after a cursory look, I don't think this line you commented on can be avoided or changed safely (maybe @mjcarroll will disagree).

Reply:
> And by the way, changing the wait set never happens with single threaded because a callback group's entities never have to be removed from the wait set since it's either executing or waiting, but never both. Which obviously isn't true for the multi-threaded executors.

I am fully aware of this, this is my main point the whole time. This also has an impact on the execution order of entities.

Reply:
> > The rclcpp::WaitSet has no concept of callback groups.
> 
> I think this might be the most important point @mjcarroll made. It's up to the executor to manipulate the wait set as needed to control what is waited on according to constraints by the callback groups.

Let's rephrase this. The rclcpp::WaitSet is the central piece of the Executor. Currently I would say, that neither the  rclcpp::Executor nor the rclcpp::WaitSet have a real concept of the callback groups. 
There is no API, to add or remove CallbackGroups in a efficient manor from the next wait. There is no way of getting a next ready executable, without iterating over X entries, that belong to a blocked callback group.

IMHO the API of rclcpp::WaitSet and rclcpp::WaitResult classes should reflect that callback groups exist, and should provide efficient ways to handle blocked callback groups. Without this, you will run into the same problems in the Multi-threaded case again and again.

And from my point of view this is a big design flaw of this PR.

Reply:
> that neither the rclcpp::Executor nor the rclcpp::WaitSet have a real concept of the callback groups.

The executors definitely do.

> IMHO the API of rclcpp::WaitSet and rclcpp::WaitResult classes should reflect that callback groups exist, and should provide efficient ways to handle blocked callback groups. Without this, you will run into the same problems in the Multi-threaded case again and again.

I have to very much disagree with this. The wait set should not have any connection to callback groups in my opinion. It's a building block that should be useful even without callbacks being used at all, if the user wants to do that. In fact that was the use case for which it was originally designed (we had users that wanted to handle their own execution and did not want to use the executor pattern).